codex-rs/core/src/codex.rs

```diff
@@ -869,9 +981,36 @@ async fn run_task(sess: Arc<Session>, sub_id: String, input: Vec<InputItem>) {
         }
     }
     sess.remove_task(&sub_id);
+
+    // Calculate total cost for OpenAI models if possible.
+    let is_openai_provider = sess
+        .client
+        .provider()
+        .base_url
+        .contains("openai.com");
+
+    let (total_cost_opt, prompt_tokens_opt, completion_tokens_opt) = if is_openai_provider {
+        let model = sess.client.model_name();
+        let (prompt_rate, completion_rate) = get_openai_pricing(model).unwrap_or((0.0, 0.0));
+        // Rates are per-token. Multiply directly.
+        let cost = (total_prompt_tokens as f64) * prompt_rate
+            + (total_completion_tokens as f64) * completion_rate;
+        (
+            Some(cost),
+            Some(total_prompt_tokens as u32),
+            Some(total_completion_tokens as u32),
+        )
+    } else {
+        (None, None, None)
+    };
```

Comment:
Can we move this into the client.rs / streaming protocol instead? Ideally we don't count or manage tokens at a *provider-level* here imho